docs/source/features/metrics.md

```diff
@@ -0,0 +1,62 @@
+
+To set up Apollo Server with Engine, [click here](https://engine.apollographql.com/) to get an Engine API key and provide it to the `ENGINE_API_KEY` environment variable. Setting an environment variable can be done in commandline(`ENGINE_API_KEY=REPLACE_WITH_YOUR_API_KEY node start-server.js`) or with the [dotenv npm package](https://www.npmjs.com/package/dotenv).
+
+## Logging
+
+Apollo Server provides two ways to log a server: per input, response, and errors or periodically throughout a request's lifecycle. Treating the GraphQL execution as a black box by logging the inputs and outputs of the system allows developers to diagnose issues quickly without being mired by lower level logs. Once a problem has been found at a high level, the lower level logs enable accurate tracing of how a request was handled.
+
+### High Level Logging
+
+To log the inputs, response, and request, Apollo Server provides three methods: `formatParams`, `formatError`, and `formatResponse`. This example uses `console.log` to record the information, servers can use other more sophisticated tools.
+
+```js
+const server = new ApolloServer({
+  typeDefs,
+  resolvers,
+  formatParams: params => {
+    console.log(params);
+    return params;
+  },
+  formatError: error => {
+    console.log(error);
+    return error;
+  },
+  formatResponse: response => {
+    console.log(response);
+    return response;
+  },
+});
+
+server.listen().then(({ url }) => {
+  console.log(`🚀  Server ready at ${url}`);
+});
+```
+
+### Granular Logs
+
+Additionally for more advanced cases, Apollo Server accepts an array of `graphql-extensions` to the `extensions` field. These extensions receive a variety of lifecycle calls for each phase of a GraphQL request and can keep state, such as the request headers. For example the `logFunction` from Apollo Server can be implemented as an [extension](https://github.com/apollographql/apollo-server/blob/8914b135df9840051fe81cc9224b444cfc5b61ab/packages/apollo-server-core/src/logging.ts) and could be modified to add additional state or functionality.
+
+```js
+const { ApolloServer }  = require('apollo-server');
+const LoggingExtension = require('./logging');
+
+const server = new ApolloServer({
+  typeDefs,
+  resolvers,
+  extensions: [() => new LoggingExtension()]
```

Comment:
Quick question: After accepting `LoggingExtension`, where will the `extensions` field be useful outside the Apollo Server constructor? Where can it be accessed outside the constructor? or does it just simply modify state?